python/lsst/cp/pipe/cpTask.py

```diff
@@ -0,0 +1,594 @@
+        # Run the CTE task
+        if self.config.doCTE:
+            self.log.info("Starting CTE task")
+            cteTaskDataId = {'run': run, 'testType': 'SFLAT_500', 'imageType': 'FLAT'}
+            for ccd in ccds:
+                if 'SFLAT_500' not in testTypes:
+                    msg = "No superflats found. Available data: %s"%testTypes
+                    if self.config.requireAllEOTests:
+                        raise RuntimeError(msg)
+                    else:
+                        self.log.warn(msg + "\nSkipping CTE task")
+                        break
+                sflatFilenames = [butler.get('raw_filename', dataId={'visit': visit,
+                                                                     'ccd': ccd})[0][:-3]
+                                  for visit in butler.queryMetadata('raw', ['visit'], dataId=cteTaskDataId)]
+                self.log.trace("CTETask: Processing %s with %s files"%(ccd, len(sflatFilenames)))
+                maskFiles = self._getMaskFiles(self.config.eotestOutputPath, ccd)
+                self.cte.run(sensor_id=ccd, superflat_files=sflatFilenames, mask_files=maskFiles)
+            del cteTaskDataId
+
+        ################################
+        ################################
+        # Run the flatPair task
+        if self.config.doFlatPair:
+            self.log.info("Starting flatPair task")
+            flatPairDataId = {'run': run, 'testType': 'FLAT', 'imageType': 'FLAT'}
+            for ccd in ccds:
+                if 'FLAT' not in testTypes:
+                    msg = "No dataset for flat_pairs found. Available data: %s"%testTypes
+                    if self.config.requireAllEOTests:
+                        raise RuntimeError(msg)
+                    else:
+                        self.log.warn(msg + "\nSkipping flatPair task")
+                        break
+                flatPairFilenames = [butler.get('raw_filename', dataId={'visit': visit,
+                                                'ccd': ccd})[0][:-3]
+                                     for visit in butler.queryMetadata('raw', ['visit'],
+                                                                       dataId=flatPairDataId)]
+                # Note that eotest needs the original filenames as that is the only place the flat
+                # pair-number is kept, so we have to resolve links and pass in the *original* paths here :(
+                # Also, there is no "flat-pair" test type, so all FLAT/FLAT imType/testType will appear here
+                # so we need to filter these for only the pair acquisitions (as the eotest code looks like it
+                # isn't totally thorough on rejecting the wrong types of data here)
+                # TODO: adding a translator to obs_comCam and ingesting this would allow this to be done
+                # by the butler instead of here.
+                flatPairFilenames = [os.path.realpath(_) for _ in flatPairFilenames if
```

Comment:
Don't use `_` as a bona fide variable name (it's used for throwaway variables).

Reply:
Apologies, it's a Robertism I caught but which I quite like(d). Very useful in py2 to prevent accidental use (due to scope-leakage), but fully pointless in py3 🙂